src/classes/HouseholdAccounts.cls

```diff
@@ -98,4 +157,362 @@ public inherited sharing class HouseholdAccounts {
     }
+
+    /*******************************************************************************************************
+    * @description Updates the accounts for contacts that have been deleted if they have other hh members
+    */
+    public void handleContactDeletion(TDTM_Runnable.DmlWrapper dmlWrapper) {
+        Set<Id> householdIdsToUpdate = new Set<Id>();
+
+        for (Account accountRecord : (List<Account>) getRecords()) {
+            Boolean needsAccountUpdate =
+                    (!accountRecord.Contacts.isEmpty()
+                            || !accountRecord.Opportunities.isEmpty());
+
+            if (needsAccountUpdate) {
+                householdIdsToUpdate.add(accountRecord.Id);
+            }
+        }
+
+        updateHouseholds(new List<Id>(householdIdsToUpdate), dmlWrapper);
+    }
+
+    /*******************************************************************************************************
+    * @description Updates an HH Account's naming, primary contact, and rollups
+    * @param accountIds the list of Account (Id's) that need updating.
+    */
+    public static void updateHouseholds(List<Id> accountIds, TDTM_Runnable.DmlWrapper dmlWrapper) {
+        if (accountIds.isEmpty()) {
+            return;
+        }
+
+        resetPrimaryContactForAccount(accountIds, dmlWrapper);
+        renameHHAccounts(accountIds);
+        rollupAccounts(accountIds);
+    }
+
+    /*******************************************************************************************************
+    * @description for the list of Household or Account Ids, update their Household Names
+    * @param accountIds the list of Household or Account Ids that need updating.
+    * @return void
+    */
+    public static void renameHHAccounts(List<Id> accountIds) {
+        if (accountIds.isEmpty()) {
+            return;
+        }
+
+        if (shouldProcessSynchronously(accountIds)) {
+            HouseholdNamingService householdNamingService = new HouseholdNamingService();
+            householdNamingService.updateHouseholdNameAndMemberCount(accountIds);
+        } else {
+            householdNamingService.updateHouseholdNameAndMemberCountAsync(accountIds);
+        }
+    }
+
+    private static Boolean shouldProcessSynchronously(List<Id> accountIds) {
+        return accountIds.size() == 1 || System.isFuture() || System.isBatch() || System.isQueueable();
+    }
+
+    /*******************************************************************************************************
+    * @description sets the Primary Account for each Account in the list of Accounts
+    * @param accountIds the list of Account (Id's) that need updating.
+    */
+    private static void resetPrimaryContactForAccount(List<Id>
+            accountIds, TDTM_Runnable.DmlWrapper dmlWrapper) {
+
+        // for each account, we need a set of their contacts.
+        Map<Id, Set<Id>> contactIdsByAccountId = new Map<Id, Set<Id>>();
+
+        for (Contact contactRecord : [
```

Comment:
Any chance we already have these Contacts queried? Feels risky to run a query (2 actually...) each time this method is called.